resources/testharness.js

```diff
@@ -2750,17 +2751,27 @@
         var remaining = Math.ceil(timeout_full / interval);
         var test_this = this;
 
-        var wait_for_inner = test_this.step_func(() => {
-            if (cond()) {
+        function step(result) {
+            if (result) {
                 func();
             } else {
-                if(remaining === 0) {
+                if (remaining === 0) {
                     assert(false, "step_wait_func", description,
                            "Timed out waiting on condition");
                 }
                 remaining--;
                 setTimeout(wait_for_inner, interval);
             }
+        }
+
+        var wait_for_inner = test_this.step_func(() => {
+            const result = cond();
+            if (result instanceof Promise) {
```

Comment:
We should avoid using `instanceof` to check for types because it fails in cross-realm cases. You could perhaps just use `Promise.resolve` here to convert the return type into a promise.

Reply:
I changed my mind, and It's using `Promise.resolve` now.